Doc/git guidelines

signekb opened the pull request:
I have started to write up some entries to explicitly state how we use/want to use Git and GitHub. I have written these more like "a series of short lessons that we could also use to teach people at Steno and others" based on Luke's feedback. 

This has resulted in the following structure:

![image](https://github.com/seedcase-project/team/assets/40836345/87911025-fb28-42a6-9c18-67e2123d082f)
[link to tree](https://tree.nathanfriend.io/?s=(%27optUs!(%27fancyj~fullPath!false~MailCSlashj~rootDotj)~source!(%27source!%27enMiK4gL-workQ-guidelYK*3CI3%20namC2Fmuch%20a%203WcoverV89sIatomic%209s29%20mKsagKV8gL0YMOtOgL%2FversU%20conMol2reposLoriK0local%20vs%20remote%7DEFGL%20zs%20filKEfil5changK8YdexIYMoductUEGLHub%20Q8_KIFtOwrL5gooB_K2YcludC%20_5teJatK2perhaps%3F8prsIFtOmak5a%20gooBpr2reviewVqimagK%206ng6RTZ%20N%20wLh%203K83C-siJeX3C-siJe.pRsiJ53C%20exaJe89sX9s.puml%207ng7RunzeB-%20modifieB-%20staged8AXA.pRsiJ5TgLhub%20Qqwhy-A*YdexIdecisU%20post%3BWmayb5b5locateBsomewher5else%3Fq...%27)~versU!%271%27)*4%20%200%20%7B2%2C%203branch4%5Cn%20%205e%206*3C-Z-N.p7*file-zC-Y-gL.p8%7D*9commLAgLhub-QBd%20CYgE%20%2B%20Fhow%20I.md0JmplKesLitMtrNdevelopmentOo%20QflowRuml0TexaJ5of%20UionV2etc.W%20shoulBX.png*YinZparallel_issuj!Mueq%7D4zMack%01zqj_ZYXWVUTRQONMLKJIFECBA987654320*)

**Most of them are ready for feedback** - excluding prs.md and issues.md.
So, please read through them and give feedback on whatever you find to be unclear or whether something is too elaborate or not elaborate enough. Thanks! :)

Note: These guidelines should probably be located somewhere else. Same goes for the GitHub-flow decision post.

signekb committed:
github flow decision entry

signekb committed:
add puml for example of github flow

signekb committed:
add draft of introduction to git workflow guidelines

signekb committed:
add drafts/outlines of additional git guideline entries

signekb committed:
add link for tilburgsciencehub source to use later

signekb committed:
change github flow figure naming to kebab-case

signekb committed:
add tilburg til reference list

signekb committed:
first draft of branches entry

signekb committed:
rename "branches.md" to "branching.md"

signekb committed:
add lowercase alphanumeric and hyphen separated to branch naming

signekb committed:
add point about consistency in branch naming

signekb committed:
add section w panel tab on branching in practice

signekb committed:
add doc in branch prefix table

signekb committed:
add ref for branch naming convention cheatsheet

signekb committed:
fix: typos

signekb committed:
docs: add introduction to commit guidelines

signekb committed:
docs: add section introducting commits

signekb committed:
docs: add how git tracks changes to aid the understanding of the  two phases of commits

signekb committed:
docs: add section on atomic commits

signekb committed:
docs: add section on commit messages

signekb committed:
docs: add section on commits in practice

signekb committed:
docs: add file-tracking-in-git figure

signekb committed:
fix: add delete new branch after merge to figures

signekb committed:
fix: add commits puml and png

signekb committed:
docs: add section with example of commit workflow

signekb committed:
docs: begin draft of issues post

lwjohnst86 commented on entries/git-workflow-guidelines/branching.md:
```diff
@@ -0,0 +1,143 @@
+| Bugfix | For fixing a bug | bugfix/(short-description) | bugfix/fix-data-cleaning-error |
+| Feature | For adding, removing, or modifying a feature* | feature/(short-description) | feature/add-variance-plot |
+| Hotfix | For quickly fixing critical issues, usually with a temporary solution | hotfix/(short-description) | hotfix/fix-select-button |
+| Doc | For writing, updating, or fixing documentation | doc/(short-description) | doc/git-guidelines |
```
I think we need more "Types" for non-code items, like the documentation and eventual training material. `doc/` is too broad.

signekb replied:
Agreed! I'll look into conventions for that

signekb replied:
@lwjohnst86 I found it difficult to find actual conventions for dividing doc into more types, but we could create our own? Something like this maybe: 

- docs-training/ or docs-learning/ (training materials) 
- docs-why/ (decision posts)
- docs-design/ (considerations on design) 
- docs-code/ (documentation on specific code?)  
- docs-blogs/ (for informal "learning"/something we have learned e.g. at a conference or meeting??)
- docs-misc/ (for things that doesn't fit any of the types we define - we can evaluate on this later to see if there are some "types" in here that we have missed) 

Do you have ideas?

signekb replied:
PS the VS Code extension [Conventional branch](https://marketplace.visualstudio.com/items?itemName=pshaddel.conventional-branch) makes it easy to create custom branch types to choose from - in a similar manner as Conventional Commits. Conventional branch is not widely used, though (391 installations)

K-Beicher replied:
I'm happy to trial it if you guys feel like we should explore it.

signekb committed:
docs: add tip on conventional commit extension for vs code

K-Beicher commented on entries/git-workflow-guidelines/index.md:
```diff
@@ -0,0 +1,79 @@
+
+### Introduction
+
+Note: In this and the following Git workflow posts, we assume that you know the basics of what Git is. If this is not the case, the Git-scm website includes great documentation, videos to get you started, as well as cheat sheets. (from VS code post)
```
Maybe a link to the Git+scm website?  Also, did you mean to leave the comment in () in there?

K-Beicher commented on entries/git-workflow-guidelines/index.md:
```diff
@@ -0,0 +1,79 @@
+- **The main branch contains production-ready code**: In this workflow, branches are created to work on new features, bug fixes, or the like and merged into the main branch as soon as the work is completed and has been reviewed. After the merge with main, the branch is deleted.
+- **One branch for one purpose**: Create a branch for one specific purpose. This allows your collaborators to easily get an overview of the ongoing work. Furthermore, the shorter the changes, the faster it will be reviewed and merged to main.
+- **Collaboration happens in the PRs**: Better work is created when you are not working in a silo. During PRs, you'll have a conversations about the proposed changes and your collaborators might have some ideas for improvement. PRs is a chance to utilise each other's experiences and expertise
+- **The change does not have to be fully implemented before creating a PR**: When you create a PR, you communicate to your collaborators what you are working on. In this [talk](https://www.youtube.com/watch?v=vCwuZfK0VG4), it is recommended to create your PRs early and not wait until your work is complete, for two reasons: 1) it helps prevent that someone else does not start doing the same work a ...
```
you have a double-negative in point 1.  prevent that they don't start

signekb replied:
Nice catch. Thanks!

K-Beicher commented on entries/git-workflow-guidelines/issues.md:
```diff
@@ -0,0 +1,52 @@
+
+One way to structure an issue is in the form of a [user story](https://www.atlassian.com/agile/project-management/user-stories). When writing the issue, this structure helps with explicitly 1) keeping the user in mind, 2) explicitly stating the functionality of the change requested, and 3) how this will benefit this kind of user.
+
+The structure is like so:
```
Suggestion: The structure looks like this:

K-Beicher commented on entries/git-workflow-guidelines/prs.md:
```diff
@@ -0,0 +1,70 @@
+- The shorter the changes, the faster it will be reviewed
+  - Also: Create your PRs early (don't wait until your feature is complete bc then someone else might start working on it) - and then you will have something to talk about
+  - Change something to talk about and then send the PR
+- CI pipeline; tests run automatically - in the PR (so they will be a part of the story of that feature) - to ensure that your PR is merged with confidence
```
I think this is the first time we introduce the concept of CI, maybe write it out here, with  (CI) afterwards, so that we are 100% sure we are all talking about the same thing (and not for instance Chief Investigator) :-)

signekb replied:
I'll remove the mention of CI pipelines here. I think it's beyond the scope of the post.

signekb committed:
docs: add note to self about focusing on why and not what in commit messages

signekb committed:
docs: :sparkles: initialise intro to git

signekb committed:
feat: :sparkles: add order to yaml header

This way, the pages will be shown in the correct order, rather than alphabetical

signekb committed:
docs: pr -> pull request

signekb committed:
docs: dev -> developer + variance -> scatter + error -> function

Minor changes to text in plot to create clearer examples

signekb committed:
docs: feature branch -> new branch

To generalise beyond adding/enhancing a new feature

signekb committed:
docs: files -> changes

To cover that commits doesn't have to be on the file level

signekb committed:
docs: update puml pngs

signekb committed:
docs: rewrite intro to avoid mentioning Seedcase and make it more learning module-like

signekb committed:
docs: add link to git.md in intro

signekb committed:
docs: self-review

signekb commented on entries/git-workflow-guidelines/branching.md:
```diff
@@ -0,0 +1,146 @@
+1. It enables you to give the branch a short, descriptive name that clearly communicates to your collaborators what work has been done on this particular branch.
+2. Smaller changes allows for a faster reviewing and merging to the main branch, since it is easier for your collaborators to get an overview of the work you have done.
+
+As an example, a branch could cover a fix to a particular bug or [issue](issues.md) or add a new feature, such as a new plot, like in the figure showing parallel development above.
```
First mention of issues. When should they be "introduced"?

K-Beicher replied:
Hmm...  Isn't a reference to a file that details them specifically a sort of intro?  Or have I misunderstood something with regards to the link?

signekb committed:
docs: change enumeration to follow order of posts

signekb committed:
docs: self-review

signekb commented on entries/git-workflow-guidelines/commits.md:
```diff
@@ -0,0 +1,213 @@
+But first, let's back up a little and look at the two phases of committing: When you want to commit some changes you have made, you first have to tell Git what changes you want to include in your commit (i.e., want to capture in the snapshot). This is done by adding them to what is known as the "staging area". Then, you save the snapshot by *committing* the staged changes with a message describin ...
+
+To understand these two phases, it might help to briefly go through how Git tracks changes (:warning: too much detail? :warning:). In the figure below (inspired by [this](https://learn.microsoft.com/en-us/azure/devops/repos/git/commits?view=azure-devops&tabs=visual-studio-2022#how-git-tracks-changes) post), it's visualised how Git tracks changes in three categories: **Unmodified**, **modified**,  ...
+
+![How Git tracks file changes](../images/file-tracking-in-git.png)
+
+*Unmodified* files includes all tracked files that haven't been created or modified since the last commit. As soon as you save a new file or save an edited file, Git tracks that as being *modified*. When you are happy with your changes, you can *stage* the modifications by adding them to the *staging area*, which includes all changes that will be a part of the next commit. Adding changes to the s ...
+
+Note that you can add as many files and changes to the staging area as you want before you commit them. All changes in the staging area at the time of commit will be a part of the same commit. In this way, the staging area allows you to control which changes you want to include in a specific commit, so you can collect the changes connected to the same unit of work in one commit (we will return to ...
+
+Commits are always created locally in the specific branch you are currently on. Therefore, it is important to check that you are on the right branch before you commit (check the [Branching](branching.md#branching-in-practice) post for how you check which branch you are currently on and how to change branch). After you have created your commits, they needs to be *pushed* to the remote repository ( ...
+
+After you have pushed your commits, your work will be backed up on the remote storage. As your work progresses, you continuously commit and push changes to the branch you are working on until you are ready to ask for feedback through a pull request (more on pull requests and the review process in the [Pull Request](prs.md) post).
```
I feel like this section is important, but I'm not sure where it should be located. After writing this, I added the "Introduction to Git" entry and maybe it would make more sense to add it there? .. But then commits will need to be introduced in that section instead? What do you think?

K-Beicher replied:
I'm not sure I would move it out, it flows quite nicely seen from a sort-of newbie's perspective

signekb replied:
Alrighty. Keep for now :-)

signekb commented on entries/git-workflow-guidelines/commits.md:
```diff
@@ -0,0 +1,213 @@
+
+Now you should have an understanding of that commits are, but how often should you commit, and how do you choose what to include in a commit?
+
+Ideally, you should commit often, based around isolated, complete changes, a practice called **atomic** commits. This means that each commit should document a *single, complete unit of work* that represents a specific idea (as described in the post [Make Atomic Git Commits](https://www.aleksandrhovhannisyan.com/blog/atomic-git-commits/#atomic-commits-and-the-single-responsibility-principle) and t ...
```
move to a reference list in the bottom of the post?

K-Beicher replied:
I personally prefer them in the text itself.

signekb replied:
Also written out like this? I could refer to them with numbers in the text, like I have done in the github flow decision post (see screenshot below)? 
![Screenshot 2023-11-21 at 10 16 50](https://github.com/seedcase-project/team/assets/40836345/49da5435-345d-448b-8f7b-08e78ce57408)

And then it looks like this in the bottom of the page:
![Screenshot 2023-11-21 at 10 17 49](https://github.com/seedcase-project/team/assets/40836345/1c887466-7a97-4a50-9848-e56a9943e87c)

signekb committed:
docs: change emphasis from italics to bold or in-line code

signekb commented on entries/git-workflow-guidelines/commits.md:
```diff
@@ -0,0 +1,213 @@
+Note that you can add as many files and changes to the staging area as you want before you commit them. All changes in the staging area at the time of commit will be a part of the same commit. In this way, the staging area allows you to control which changes you want to include in a specific commit, so you can collect the changes connected to the same unit of work in one commit (we will return to ...
+
+Commits are always created locally in the specific branch you are currently on. Therefore, it is important to check that you are on the right branch before you commit (check the [Branching](branching.md#branching-in-practice) post for how you check which branch you are currently on and how to change branch). After you have created your commits, they needs to be *pushed* to the remote repository ( ...
+
```
pushing and prs might not be relevant here? remove? or is it nice with a reminder that you have to push and get your changes reviewed?

signekb commented on entries/git-workflow-guidelines/commits.md:
```diff
@@ -0,0 +1,213 @@
+Ideally, you should commit often, based around isolated, complete changes, a practice called **atomic** commits. This means that each commit should document a *single, complete unit of work* that represents a specific idea (as described in the post [Make Atomic Git Commits](https://www.aleksandrhovhannisyan.com/blog/atomic-git-commits/#atomic-commits-and-the-single-responsibility-principle) and t ...
+
+An atomic commit can also be described as a commit of the smallest possible size, that can be undone without any unwanted side effects apart from what would be expected based on its commit message. A commit is not atomic, if undoing it removes other changes than described.
+
```
Does "the smallest possible size" sound a bit like you should commit every little change individually?

rqding replied:
I think we should be more clear about commit vs push. The one commit should be one smaller but also complete change which makes the git reverse more easier.

signekb replied:
Good point. I'll add something about that :)

signekb committed:
docs: add introduction

signekb committed:
docs: Add short description of Git + Remove notes

signekb committed:
docs: Add section on local and remote repositories

signekb committed:
docs: Add figure on local and remote repos

signekb committed:
docs: Minor text edits to What is Git section

signekb committed:
docs: Remove warnings in the beginning of the entry

signekb committed:
feat: Add location of remote/local repo

signekb committed:
docs: Revision of existing sections

signekb committed:
docs: Add summary of Git post

signekb committed:
docs: Add section on how to initialise a Git repo

signekb committed:
docs: Add section on how to clone an existing Git repo

signekb committed:
docs: Remove note since I have now added the init and clone sections

signekb committed:
docs: Add Git install and GitHub account + init and clone to summary section

K-Beicher reviewed (dismissed):
Seems sensible

signekb committed:
docs: self-review of intro to commits section

signekb committed:
docs: shorten intro in commits post

signekb committed:
docs: self-review: atomic commits section

signekb committed:
docs: self-review: commit messages section

signekb committed:
docs: self-review of commits in practice section (adding how-tos for VS code and terminal)

signekb committed:
docs: self-review  of partial commits section

signekb committed:
docs: :memo: add undoing a commit section

signekb committed:
docs: remove trailing whitespace

signekb committed:
docs: :memo: add summary section to commits

signekb committed:
docs: remove blank line

signekb committed:
Revert "docs: self-review of intro to commits section"

This reverts commit 04f17be23bdc1f7b204e1e927421215d5dcf37c4.
I accidently included _old.md, including irrelevant personal notes

signekb committed:
docs: :memo: include how to see existing branches in vs code

signekb committed:
docs: remove "commit message" from commits figure

signekb committed:
docs: add file tracking in git figure showing new and existing files

signekb committed:
docs: add alternative figure on how git tracks files (showing new and existing file tracking separately)

signekb committed:
docs: Include more text in review and suggested changes points in github flow figure

signekb committed:
docs: Include github flow figure with highlight on current part in each post

signekb committed:
docs: first draft of pull request and review intro

signekb committed:
docs: Add "reviews" to title of post to highlight the importance of the review process as well

signekb committed:
docs: first draft of introduction to prs section

signekb committed:
docs: Add that PR description should include summary of changes and, if relevant, #issue

signekb committed:
docs: :art: change pseudo code blocks including commit messages from bash to default

By changing these blocks to "default", all text in the code blocks is now the same colour. E.g., the opening paranthesis in commit messages including scope were shown in red in bash code blocks. By changing it to "default", the parenthesis is now the same colour as the rest of the text in the block.

signekb commented:
@K-Beicher @rqding @lwjohnst86 
Since this is quite a lot of text, I think (if you have time) you can start to review most of them. I am starting to feel a bit lost in all of it, so hopefully some fresh eyes could help me out a bit :) 
Please don't review prs.md and issues.md yet (since I'm currently working on the material there)

signekb committed:
docs: add jeff atwood quote

signekb committed:
docs: add review guidelines from reviewing-prs entry in community

Directly copypasted from community/guide-entries/reviewing-prs/index.qmd (https://github.com/seedcase-project/seedcase-project/blob/main/community/guide-entries/reviewing-prs/index.qmd)

signekb committed:
docs: add missing word "through"

signekb committed:
docs: add "rich diff" tip for reviewing md or qmd files

signekb committed:
docs: add merge conflict section

signekb committed:
docs: add github workflow figure with emphasis on prs and reviews to intro

signekb committed:
docs: update intro

signekb committed:
docs: update introduction to prs and reviews section

signekb committed:
docs: add tip about conventional branch extension to branching in vs code section

signekb committed:
docs: remove irrelevant text ("copy code")

signekb committed:
docs: minor text mods (congruence, etc.) i intro

signekb committed:
docs: minor text mods to summary section

signekb committed:
docs: remove simplified (and incorrect) file tracking in git figure

signekb committed:
docs: add brief "git workflows in practice" section

signekb committed:
style: remove blank line

signekb committed:
style: add `` around Terminal output

signekb committed:
docs: repository -> repository's history

signekb committed:
docs: fix broken link to git guidelines index.md

signekb committed:
docs: fix typos

signekb committed:
docs: minor text mods

signekb committed:
docs: include the benefits of issues in the introduction

signekb committed:
docs: update introduction to issues with list of central aspects

signekb committed:
docs: add section with issue types, including feature req, bug report, and user story

lwjohnst86 committed:
docs: :memo: Edits and small formatting changes to "what is git" section

lwjohnst86 committed:
Merge branch 'feature/git-guidelines' of https://github.com/seedcase-project/team into feature/git-guidelines

signekb committed:
docs: add tip about issue templates

lwjohnst86 reviewed (commented):
Just finishing this comment before the meeting! :stuck_out_tongue:

signekb committed:
docs: change codeblocks to default for nicer layout

signekb committed:
docs: add link to issue templates

signekb commented:
Moved to iteration in Feb 2024, left as In Review on the Oct 2023 iteration

signekb committed:
docs: add description of prefixes to issue types

signekb committed:
docs: add issues in practice section

signekb committed:
docs: add paneltab with issues using the command line

issues using github and issues in vs code are still wips

signekb committed:
docs: add summary section header

signekb committed:
docs: add prs in practice section

signekb committed:
docs: add tip on creating prs early and drafts

signekb committed:
docs: add tip on pr templates

signekb committed:
docs: add paneltab on creating prs (very much wip)

signekb committed:
docs: WIP add very drafty review in practice section

signekb committed:
docs: add note on fetch section

lwjohnst86 committed:
Merge branch 'feature/git-guidelines' of https://github.com/seedcase-project/team into feature/git-guidelines

lwjohnst86 committed:
docs: :pencil2: Small edits

signekb commented on entries/git-workflow-guidelines/git.md:
```diff
@@ -0,0 +1,197 @@
+- Could also include sections from [R Cubed Intro - Version Control](https://r-cubed-intro.rostools.org/sessions/version-control)
+  - Sections "6.1 What is version control?", "6.2 What is Git?", and "6.3 Basics of Git" could be relevant here
+
+## Repositories: Local and remote
```
Very true, very true! Will split it up and add the missing information when I have the time.

lwjohnst86 commented:
@signekb as we discussed, I will merge this in, move the content over into rostools as part of the course infrastructure there, and then delete everything in this repo.
